apps/machine/microblaze_generic/zynqmp_mb_a53_rproc.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2020 Xilinx, Inc.
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**************************************************************************
+ * FILE NAME
+ *
+ *       zynqmp_mb_a53_rproc.c
+ *
+ * DESCRIPTION
+ *
+ *      OpenAMP remoteproc implementation for Xilinx MicroBlaze design
+ *      example. The design instanciates Xilinx MicroBlaze soft proccesor
+ *      in ZynqMP's programmable logic and uses processing system (PS)
+ *      DDR memory for its text and data. The APU in PS runs Linux with
+ *      reserved-memory nodes in its device tree toaccommodat the
+ *      MicroBlaze text, data.
+ *
+ **************************************************************************/
+
+#include <metal/atomic.h>
+#include <metal/assert.h>
+#include <metal/device.h>
+#include <metal/irq.h>
+#include <metal/utilities.h>
+#include <openamp/rpmsg_virtio.h>
+#include "platform_info.h"
+
+static struct remoteproc *
+zynqmp_mb_a53_proc_init(struct remoteproc *rproc, struct remoteproc_ops *ops,
+			void *arg)
+{
+	struct remoteproc_priv *prproc = arg;
+
+	if (!rproc || !prproc || !ops)
+		return NULL;
+	rproc->priv = prproc;
+	rproc->ops = ops;
+	return rproc;
+}
+
+static inline void zynqmp_mb_a53_proc_remove(struct remoteproc *rproc)
+{
+	(void)rproc;
+}
+
+static void *
+zynqmp_mb_a53_proc_mmap(struct remoteproc *rproc, metal_phys_addr_t *pa,
+			metal_phys_addr_t *da, size_t size,
+			unsigned int attribute, struct metal_io_region **io)
+{
+	struct remoteproc_mem *mem;
+	metal_phys_addr_t lpa, lda;
+	struct metal_io_region *tmpio;
+
+	lpa = *pa;
+	lda = *da;
+
+	if (lpa == METAL_BAD_PHYS && lda == METAL_BAD_PHYS)
+		return NULL;
+	if (lpa == METAL_BAD_PHYS)
+		lpa = lda;
+	if (lda == METAL_BAD_PHYS)
+		lda = lpa;
+
+	mem = metal_allocate_memory(sizeof(*mem));
+	if (!mem)
+		return NULL;
+	tmpio = metal_allocate_memory(sizeof(*tmpio));
+	if (!tmpio) {
+		metal_free_memory(mem);
+		return NULL;
+	}
+	remoteproc_init_mem(mem, NULL, lpa, lda, size, tmpio);
+	/* va is the same as pa in this platform */
+	metal_io_init(tmpio, (void *)lpa, &mem->pa, size,
+		      sizeof(metal_phys_addr_t) << 3, attribute, NULL);
+	remoteproc_add_mem(rproc, mem);
+	*pa = lpa;
+	*da = lda;
+	if (io)
+		*io = tmpio;
+	return metal_io_phys_to_virt(tmpio, mem->pa);
```

Comment:
free memory on error

Reply:
what do you mean here? as this mirrors the zynqmp_r5 zynqmp_r5_a53_rproc.c implementation that is already merged

Reply:
So it  has probably  to  be fix in zynqmp_r5_a53_rproc too. :)
As this code is an example that is a reference for openamp user development, i think it important to have at least for ne dev a good management of the error case and init/deinit. This is the aim of my comments in your PR.
For instance here a memory is allocated but not free in case of error...

Reply:
ok

Reply:
https://github.com/OpenAMP/open-amp/issues/230  created